### PR TITLE
Include `cmm-sources` when linking shared objects

### LINKED ISSUE #7182: `cmm-sources` are not included in shared object linking
**Describe the bug**
`cmm-sources` objects are not linked into shared objects.

**To Reproduce**

```
$ cd Cabal/cabal-testsuite/PackageTests/CmmSources
$ cabal v2-build --enable-executable-dynamic
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - cmmexperiment-0 (exe:demo) (first run)
Preprocessing executable 'demo' for cmmexperiment-0..
Building executable 'demo' for cmmexperiment-0..
Linking /opt/exp/ghc/ghc-landing/libraries/Cabal/cabal-testsuite/PackageTests/CmmSources/dist-newstyle/build/x86_64-linux/ghc-8.8.3/cmmexperiment-0/x/demo/build/demo/demo ...
/opt/exp/ghc/ghc-landing/libraries/Cabal/cabal-testsuite/PackageTests/CmmSources/dist-newstyle/build/x86_64-linux/ghc-8.8.3/cmmexperiment-0/build/libHScmmexperiment-0-inplace-ghc8.8.3.so: error: undefined reference to 'aToMyWordzh'
collect2: error: ld returned 1 exit status
`cc' failed in phase `Linker'. (Exit code: 1)
```

Please use version-prefixed commands (e.g. `v2-build` or `v1-build`) to avoid ambiguity.

**Expected behavior**
A runnable executable should be producd.

**System information**
 - Any
 - `cabal` 5aea8a4b8463e1ae95272e190a1022764164294f

**Additional context**

### PR DESCRIPTION
**Pull request by mpickering**
Fixes #7182 

Patch is by @bgamari, I have just rebased it on master as requested by @phadej in #7183

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog (add file to `changelog.d` directory).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by mpickering** [comment]
Anyone fancy merging this patch?

**Comment by bgamari** [comment]
This should also end up in the 3.6 branch.

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 1]
```diff
@@ -1365,7 +1365,9 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
                                              [tmpDir </> x | x <- cObjs ++ cxxObjs]
                     }
       dynLinkerOpts = mempty {
-                      ghcOptRPaths         = rpaths
+                      ghcOptRPaths         = rpaths,
+                      ghcOptInputFiles     = toNubListR
+                                             [tmpDir </> x | x <- cObjs ++ cxxObjs]
```
Comment (fgaz) [inline Cabal/src/Distribution/Simple/GHC.hs:1370]: echoing @phadej's question:

> This change is something else? We don't need cmmObjs nor asmObjs however?

Comment (mpickering) [reply Cabal/src/Distribution/Simple/GHC.hs:1370]: I don't really understand the question. Before the files weren't even passed the the linker, now they are. It is possible that these other files are also needed but this patch fixes the issue for cxxObjs, as claimed, and other files types are out-of-scope.

Comment (fgaz) [reply Cabal/src/Distribution/Simple/GHC.hs:1370]: > I don't really understand the question.

Rephrasing as I understand it: Is this part of the "Include cmm-sources and asm-sources when linking shared objects" goal? If yes, why aren't cmmObjs and asmObjs included here too?

> this patch fixes the issue for cxxObjs

wasn't it cmm and asm?

Comment (mpickering) [reply Cabal/src/Distribution/Simple/GHC.hs:1370]: `cObjs` contains all the objs you list as it is constructed from cLikeFiles (which contains asm, cmm, c etc).

https://github.com/haskell/cabal/blob/5f691229abd511884ce2cca5b33736222b143000/Cabal/src/Distribution/Simple/GHC.hs#L554

**Comment by phadej** [comment]
The confusion steps from having lines

```haskell
      cObjs       = map (`replaceExtension` objExtension) cLikeFiles 
```

and

```haskell
      cObjs               = map (`replaceExtension` objExtension) cSrcs
```

Rename the first to `cLikeObjs` (and all other related "just" C) and then confusion will be no more.

In particular this is *confusing* because it seems that there is split to "C, asm, cmm" and "C++" (cxx). However `cxxSources` **are** in `cLikeFiles`.

THere is code like

```haskell
    let cProfObjs            = map (`replaceExtension` ("p_" ++ objExtension))
                               (cSources libBi ++ cxxSources libBi)
        cSharedObjs          = map (`replaceExtension` ("dyn_" ++ objExtension))
                               (cSources libBi ++ cxxSources libBi)
```

why that doesn't include `cmm` and `asm` files? They don't have profiling variants?

There are a lot of questions, and variable naming & code structure doesn't give answers.

**Comment by mpickering** [comment]
I am not sure if @phadej is requesting changes or just suggesting an improvement. We didn't modify the names in this confusing code.

This patch is critical to build ghc-debug with 9.2 so needs to be in the 3.6 branch.

**Review by emilypi — APPROVED** [review]
Thanks @mpickering  - this is clear and even tests!

**Comment by bgamari** [comment]
For what it's worth, I agree with @phadej that the current naming is confusing; I struggled to figure out its semantics when I first wrote the patch as well. It would be quite reasonable to carry out @phadej's proposed renaming in a follow-up patch.

**Comment by mpickering** [comment]
Anyone want to merge this?

**Comment by emilypi** [comment]
As long as we're agreed there will be a follow up for @phadej's commentary, yes @mpickering - consider it merged

**Comment by phadej** [comment]
Can we have an issue for

> I struggled to figure out its semantics when I first wrote the patch as well. It would be quite reasonable to carry out @phadej's proposed renaming in a follow-up patch.

Just to remind us when this code is touched another time.

**Comment by bgamari** [comment]
@emilypi, can you also backport to the 3.6 branch?

**Comment by fgaz** [comment]
@bgamari There isn't a 3.6 branch yet

**Comment by fgaz** [comment]
>     * [x]  Any changes that could be relevant to users have been recorded in the changelog (add file to `changelog.d` directory).

The changelog entry isn't there. @mpickering or @bgamari could you add one in a follow-up pr?

**Comment by bgamari** [comment]
> @bgamari There is no 3.6 branch yet

Ahh, I see. Cabal hasn't forked yet; support for GHC 9.2 has merely been added.

> The changelog entry isn't there. @mpickering or @bgamari could you add one in a follow-up pr?

Sure.

**Pull request merged**